xds/src/main/java/io/grpc/xds/Matcher.java

```diff
@@ -0,0 +1,476 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+import com.google.re2j.Pattern;
+import java.net.InetAddress;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/** Defines matcher abstract and provides a group of request matchers.
+ * A matcher evaluates an {@link EvaluateArgs} input and tells whether certain
+ * argument in the input matches a predefined matching pattern. */
+public abstract class Matcher {
+  protected Matcher() {}
+
+  public abstract boolean matches(EvaluateArgs args);
+
+  /** Matches when any of the matcher matches. */
+  public static class OrMatcher extends Matcher {
+    private final List<? extends Matcher> anyMatch;
+
+    public OrMatcher(List<? extends Matcher> matchers) {
+      checkNotNull(matchers, "matchers");
+      this.anyMatch = Collections.unmodifiableList(matchers);
+    }
+
+    @Override
+    public boolean matches(EvaluateArgs args) {
+      for (Matcher m : anyMatch) {
+        if (m.matches(args)) {
+          return true;
+        }
+      }
+      return false;
+    }
+  }
+
+  /** Matches when all of the matchers match. */
+  public static class AndMatcher extends Matcher {
+    private final List<? extends Matcher> allMatch;
+
+    public AndMatcher(List<? extends Matcher> matchers) {
+      checkNotNull(matchers, "matchers");
+      this.allMatch = Collections.unmodifiableList(matchers);
+    }
+
+    @Override
+    public boolean matches(EvaluateArgs args) {
+      for (Matcher m : allMatch) {
+        if (!m.matches(args)) {
+          return false;
+        }
+      }
+      return true;
+    }
+  }
+
+  /** Always true matcher.*/
+  public static class AlwaysTrueMatcher extends Matcher {
+    public static AlwaysTrueMatcher INSTANCE = new AlwaysTrueMatcher();
+
+    @Override
+    public boolean matches(EvaluateArgs args) {
+      return true;
+    }
+  }
+
+  /** Negate matcher.*/
+  public static class InvertMatcher extends Matcher {
+    private final Matcher toInvertMatcher;
+
+    public InvertMatcher(Matcher matcher) {
+      this.toInvertMatcher = matcher;
+    }
+
+    @Override
+    public boolean matches(EvaluateArgs args) {
+      return !toInvertMatcher.matches(args);
+    }
+  }
+
+  /** Matcher for HTTP request path.
+   * TODO(zivy@): merge with {@link PathMatcher} */
```

Comment:
Why isn't this done in this PR?

Reply:
RouteMatcher (btw, this name is misleading) does a subset of PathMatcher, it cannot do `suffix` and `contains` matching. It may due to historical reasons for Envoy to have two definitions for path matching. But I think we could unify them in gRPC's implementation. We could use PathMatcher for places where RouteMatchers are used, `suffix` and `contains` matching will just never be used for the routing feature (it will never be triggered as the control plane will never have `suffix` or `contains` matching configuration).

Reply:
The route path matcher has an additional [field](https://github.com/grpc/grpc-java/blob/master/xds/third_party/envoy/src/main/proto/envoy/config/route/v3/route_components.proto#L465) other than the rbac path matcher, although xds likely does not use it for now. But not sure they will further diverge in the future. That was my main point to keep them separated. But of course it looks more concise to not duplicate them.

Also even though `suffix` and `contains` likely will not be sent from Control Plane for route path matcher, it ideally still requires validation at xds client.

Reply:
> The route path matcher has an additional field other than the rbac path matcher, although xds likely does not use it for now. But not sure they will further diverge in the future. That was my main point to keep them separated. But of course it looks more concise to not duplicate them.

That's something special-casing CONNECT requests, will never likely to be used by gRPC. Actually, based on the layer of xDS path matching lies on gRPC, it will never be applicable. Also, if you are worrying about something more advanced for path matching to be added in the future, that thing should be added for all usages of path matching.

> Also even though suffix and contains likely will not be sent from Control Plane for route path matcher, it ideally still requires validation at xds client.

Why? Validation applies to the control plane messages received, which does not have `suffix` or `contains` in its schema. We are just using a big container to receive a small thing.

Reply:
SG.